### PR TITLE
Minify static content by transpiling it first

### LINKED ISSUE #409: Static js needs to be transpiled before minification
As we don't transpile to es2015 completely, uglify will trow. 
Proposed solution: swap uglify with [babili](https://github.com/babel/babili)

### PR DESCRIPTION
**Pull request by matteofigus**
Fixes #409 

Summary: #418 was opened and merge to fix #409 but we had to rollback as the packaging was producing an invalid bundle (not good for browsers and potentially fragile for uglify). So #420 reverted it. This should be finally the one, with more testing, more fixes etc.

To review, I recommend to go commit by commit.

### CODE REVIEW

File: src/cli/domain/package-static-files/minify-file.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+var babel = require('babel-core');
+var babelPresetEnv = require('babel-preset-env');
+var CleanCss = require('clean-css');
+var uglifyJs = require('uglify-js');
+
+module.exports = function(fileType, fileContent){
```
Comment (mattiaerre) [inline src/cli/domain/package-static-files/minify-file.js]: this should be `fileExt` correct?

File: test/integration/cli-domain-package-static-files-minify-file.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		describe('when file contains es6', function(){
+
+			var content = 'const hi = (name) => `hello ${name}`;';
+			var minified;
+
+			before(function(){
+				minified = minifyFile('.js', content);
+			});
+
+			it('should minify it', function(){
```
Comment (mattiaerre) [inline test/integration/cli-domain-package-static-files-minify-file.js]: a snapshot test, in this case, would simplify what you need to assert in order to make sure that the minification happened.

File: src/cli/domain/package-static-files/index.js [hunk 3]
```diff
@@ -64,7 +34,7 @@ var copyDir = function(params, cb){
 
         if(params.minify && params.ocOptions.minify !== false && (fileExt === '.js' || fileExt === '.css')){
           var fileContent = fs.readFileSync(filePath).toString(),
-              minified = minifyFile(fileExt, fileContent, params.ocOptions);
+              minified = minifyFile(fileExt, fileContent);
```
Comment (mattiaerre) [inline src/cli/domain/package-static-files/index.js]: this is the only consumer of `minifyFile` I presume.

File: src/cli/domain/package-static-files/minify-file.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        es5 = babel.transform(fileContent, babelOptions).code;
+    
+    return uglifyJs.minify(es5, { fromString: true }).code;
+
+  } else if(fileType === '.css'){
+
+    return new CleanCss().minify(fileContent).styles;
+  }
+
+  return fileContent;
```
Comment (mattiaerre) [inline src/cli/domain/package-static-files/minify-file.js]: you shouldn't need to return here as the file extension check is done also at a higher level.

File: test/unit/cli-domain-package-components.js [hunk 5]
```diff
@@ -6,96 +6,146 @@ var path = require('path');
...
+      join: path.join,
+      resolve: function(){
+        return _.toArray(arguments).join('/');
       }
-    },
-    path: pathMock,
-    './package-static-files': sinon.stub().yields(null, 'ok'),
-    './package-template': sinon.stub().yields(null, { type: 'jade', src: 'template.js', hashKey: '123456'})
-  }, { __dirname: '' });
+    };
```
Comment (mattiaerre) [inline test/unit/cli-domain-package-components.js]: I see a lot of `arrange` before running this unit test. maybe this component does too many things?

File: test/unit/cli-domain-package-static-files.js [hunk 6]
```diff
@@ -6,69 +6,75 @@ var path = require('path');
...
-    'node-dir': { paths: sinon.stub().yields(null, { files:[] })},
-    path: {
-      basename: path.basename,
-      dirname: function(){
-        return path.dirname.apply(this, _.toArray(arguments)).replace(/\\/g, '/');
-      },
-      extname: path.extname,
-      join: function(){
-        return path.join.apply(this, _.toArray(arguments)).replace(/\\/g, '/');
+describe('cli : domain : package-static-files', function(){
```
Comment (mattiaerre) [inline test/unit/cli-domain-package-static-files.js]: same as above re the `arrange` phase.

File: test/unit/cli-domain-package-components.js (hunk 5 above)
Comment (matteofigus) [inline test/unit/cli-domain-package-components.js]: Agree. But I would separate in a different PR if you don't mind

File: test/unit/cli-domain-package-static-files.js (hunk 6 above)
Comment (matteofigus) [inline test/unit/cli-domain-package-static-files.js]: Agree. That's why I am starting to take one piece off (minify part). Definitely gonna tackle other little pieces but I would do in separate PRs if you don't mind

File: src/cli/domain/package-static-files/index.js (hunk 3 above)
Comment (matteofigus) [inline src/cli/domain/package-static-files/index.js]: Yes

File: test/integration/cli-domain-package-static-files-minify-file.js (hunk 2 above)
Comment (matteofigus) [inline test/integration/cli-domain-package-static-files-minify-file.js]: With a snapshot I would test that
```js
const hi = (name) => `hello ${name}`;
```
would always translate to
```js
var hi=function(n){return"hello "+n};
```
Which is fine in this scenario, but what if at some point I upgrade babel and this starts translating to 
```js
var hi=function(a){return"hello "+a};
```
So, my thinking was that I wanted to test something without taking the risk of coupling it too much to the snapshot. So I guess that was by design. Perhaps I'm overthinking this - as I see the benefit of seeing the simplicity of A to B transformation for a test readability point of view.

Comment (nickbalestra) [inline test/integration/cli-domain-package-static-files-minify-file.js]: I think that the process is referentially transparent , so it should return always the same result

Comment (matteofigus) [inline test/integration/cli-domain-package-static-files-minify-file.js]: ok, so let me update this.

File: test/unit/cli-domain-package-components.js (hunk 5 above)
Comment (mattiaerre) [inline test/unit/cli-domain-package-components.js]: Sure, happy to pair on this if you like.

File: test/integration/cli-domain-package-static-files-minify-file.js (hunk 2 above)
Comment (mattiaerre) [inline test/integration/cli-domain-package-static-files-minify-file.js]: I personally think we should not test how the internal minification library works; that's why I was proposing a snapshot test. You will end up having a file containing your uglified code in one place. which is more or less what you do have now. What about a pairing session w/ you after this PR has been merged?

// cc @matteofigus @nickbalestra

**Comment by mattiaerre** [comment]
@matteofigus this LGTM, happy to merge if you like.
